pkg/clustermanager/provider.go

```diff
@@ -0,0 +1,231 @@
+		if ok {
+			discoveryPrefix = providerDesc.Spec.ClusterPrefixForDiscovery
+			if discoveryPrefix == "" {
+				discoveryPrefix = "*"
+			}
+		}
+	}
+
+	p := &provider{
+		name:            providerName,
+		providerClient:  newProviderClient,
+		c:               c,
+		nameSpace:       ProviderNS(providerName),
+		discoveryPrefix: discoveryPrefix,
+	}
+
+	c.providers[providerName] = p
+	return p, nil
+}
+
+func (p *provider) filterOut(lcName string) bool {
+	if p.discoveryPrefix == "*" {
+		return false
+	}
+	return !strings.HasPrefix(lcName, p.discoveryPrefix)
+}
+
+// StartDiscovery will start watching provider clusters for changes
+func (p *provider) StartDiscovery() error {
+	watcher, err := p.providerClient.Watch()
+	if err != nil {
+		return err
+	}
+	p.providerWatcher = watcher
+	go p.processProviderWatchEvents()
+
+	return nil
+}
+
+// StopDiscovery will stop watching provider clusters for changes
+func (p *provider) StopDiscovery() error {
+	p.c.lock.Lock()
+	defer p.c.lock.Unlock()
+
+	if p.providerWatcher == nil {
+		return errors.New("failed to stop provider discovery. provider watcher does not exist")
+	}
+	p.providerWatcher.Stop()
+	return nil
+}
+
+func (p *provider) processProviderWatchEvents() {
+	ctx := context.Background()
+	logger := klog.FromContext(ctx)
+	var reflcluster *lcv1alpha1apis.LogicalCluster
+	for {
+		event, ok := <-p.providerWatcher.ResultChan()
+		if !ok {
+			logger.Info("Cluster provider watch was stopped", "provider", p.name)
+			return
+		}
+		lcName := event.Name
+		reflclusterObj, found, errLC := p.c.logicalClusterInformer.GetIndexer().GetByKey(lcKeyFunc(p.nameSpace, lcName))
+
+		if found {
+			reflcluster, ok = reflclusterObj.(*lcv1alpha1apis.LogicalCluster)
+			if !ok {
+				runtime.HandleError(errors.New("unexpected object type. expected ClusterProviderDesc"))
+				continue
+			}
+		}
+
+		if !found || (found && !reflcluster.Spec.Managed) {
+			// For unmanaged clusters discover & ypdate only clusters that match the provider prefix
+			if p.filterOut(lcName) {
+				continue
+			}
+		}
+		switch event.Type {
+		case watch.Added:
+			logger.Info("New cluster was detected", "cluster", event.Name, "provider", p.name)
+			// A new cluster was detected either create it or change the status to READY
+			if !found || errLC != nil {
+				// No corresponding Logicalcluster, let's create it
+				logger.V(2).Info("Creating new LogicalCluster object", "cluster", event.Name)
+				lcluster := lcv1alpha1apis.LogicalCluster{}
+				lcluster.Name = lcName
+				lcluster.Spec.ClusterProviderDescName = p.name
+				lcluster.Spec.Managed = false
+				lcluster.Status.ClusterConfig = event.LCInfo.Config
+				lcluster.Status.Phase = lcv1alpha1apis.LogicalClusterPhaseReady
+				_, err := p.c.clientset.LogicalclusterV1alpha1().LogicalClusters(p.nameSpace).Create(ctx, &lcluster, v1.CreateOptions{})
+				chkErrAndReturn(logger, err, "Detected New cluster. Couldn't create the corresponding LogicalCluster", "cluster name", lcName)
```

Comment:
Rather than forget about the error, would it be better to use the controller pattern here, so that the cluster reference could be requeued for a later retry?